pcsx2-qt/Debugger/RegisterWidget.cpp

```diff
@@ -0,0 +1,451 @@
+ *  of the GNU Lesser General Public License as published by the Free Software Found-
+ *  ation, either version 3 of the License, or (at your option) any later version.
+ *
+ *  PCSX2 is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+ *  PURPOSE.  See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with PCSX2.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "PrecompiledHeader.h"
+
+#include "RegisterWidget.h"
+
+#include "common/BitCast.h"
+
+#include "QtUtils.h"
+#include <QtGui/QMouseEvent>
+#include <QtWidgets/QTabBar>
+#include <QtWidgets/QStylePainter>
+#include <QtWidgets/QStyleOptionTab>
+#include <QtGui/QClipboard>
+#include <QtWidgets/QInputDialog>
+#include <QtWidgets/QProxyStyle>
+#include <QtWidgets/QMessageBox>
+
+#include <algorithm>
+
+#define CAT_SHOW_FLOAT (categoryIndex == EECAT_FPR && m_showFPRFloat) || (categoryIndex == EECAT_VU0F && m_showVU0FFloat)
+
+using namespace QtUtils;
+
+RegisterWidget::RegisterWidget(QWidget* parent)
+	: QWidget(parent)
+	, m_palette(QApplication::palette())
+{
+	this->setContextMenuPolicy(Qt::ContextMenuPolicy::CustomContextMenu);
+
+	ui.setupUi(this);
+	ui.registerTabs->setDrawBase(false);
+
+	connect(this, SIGNAL(customContextMenuRequested(QPoint)), SLOT(customMenuRequested(QPoint)));
+	connect(ui.registerTabs, SIGNAL(currentChanged(int)), SLOT(tabCurrentChanged(int)));
+};
+
+RegisterWidget::~RegisterWidget()
+{
+}
+
+void RegisterWidget::SetCpu(DebugInterface* cpu)
+{
+	m_cpu = cpu;
+	for (int i = 0; i < m_cpu->getRegisterCategoryCount(); i++)
+	{
+		ui.registerTabs->addTab(m_cpu->getRegisterCategoryName(i));
+	}
+
+	connect(ui.registerTabs, &QTabBar::currentChanged, [this]() { this->repaint(); });
+}
+
+void RegisterWidget::tabCurrentChanged(int cur)
+{
+	m_rowStart = 0;
+}
+
+void RegisterWidget::paintEvent(QPaintEvent* event)
+{
+	if (!m_cpu)
+		return;
+
+	QPainter painter(this);
+	painter.setPen(m_palette.text().color());
+	m_renderStart = QPoint(0, ui.registerTabs->pos().y() + ui.registerTabs->size().height());
+	const QSize renderSize = QSize(this->size().width(), this->size().height() - ui.registerTabs->size().height());
+
+	m_rowHeight = painter.fontMetrics().height() + 2;
+	m_rowEnd = m_rowStart + (renderSize.height() / m_rowHeight) - 1; // Maybe move this to a onsize event
+
+	bool alternate = m_rowStart % 2;
+
+	const int categoryIndex = ui.registerTabs->currentIndex();
+
+	// Used for 128 bit and VU0f registers
+	const int titleStartX = m_renderStart.x() + (painter.fontMetrics().averageCharWidth() * 6);
+	m_fieldWidth = ((renderSize.width() - (painter.fontMetrics().averageCharWidth() * 6)) / 4);
+	const int fieldCharCount = std::min(((int)m_fieldWidth / painter.fontMetrics().averageCharWidth()) - 1, 8);
```

Comment:
Not used?

Reply:
Yeah there's a few unused things around I need to clean up